Add the ability to specify ffmpeg path & use Matplotlib's ffmpeg path

calw20 opened the pull request:
Changed default ffmpeg command to use matplotlib.pyplot.rcParams['animation.ffmpeg_path'] and added "ffmpeg_path" as a new argument to the Loom initialiser to allow for each Loom instance to have its own ffmpeg command path (defaults to matplotlib's).

Sorry for the double pull-request - I managed to mangle my git history.

calw20 committed:
Add the ability to specify ffmpeg path & use MatPlotLib's ffmpeg path

calw20 committed:
Add mutli-thread check to the init warning.

calw20 committed:
Whitelist file extensions and do check

calw20 committed:
Quick check to make sure frame counter is never None

calw20 committed:
Add checks to make sure Scale Filters are valid

calw20 committed:
Fix badly placed assert

ali-ramadhan commented on matplotloom/__init__.py:
```diff
@@ -1,23 +1,24 @@
 __all__ = ["Loom"]
 
 
 def _check_ffmpeg_availability():
     """Check if ffmpeg is available on the system."""
```
Should the user-supplied `ffmpeg_path` be an input here? I guess it can't be since this is `__init__.py`. I guess in the instance that ffmpeg is not available on the system but the user-supplied path works this will still produce a warning, but maybe this is fine?

calw20 replied:
Yeah that was my thought as well but then I saw it was passed to `__init__.py` I couldn't figure out a way to make it stick. The hope was that by using matplotlib's rcPrams if a user updated the rcPrams _before_ importing `Loom` it would pull the updated value.

This _works_ but only in the context of the main thread, as soon as you parallelise `Loom` it re-imports matplotlib & its rcPrams seems to be reset to the file-based defaults. I'm yet to test it but in theory if you add an rcPrams file where matplotlib expects it those settings should be maintained in the mutli-threaded/process context.

ali-ramadhan commented on matplotloom/loom.py:
```diff
@@ -10,6 +10,10 @@
 
+DEFAULT_FFMPEG_PATH = plt.rcParams['animation.ffmpeg_path']
+
+ACCEPTABLE_EXTENSIONS = ("mp4", "gif")
```
Hmmm, I could be wrong but I thought `Loom` should still work if you specify another extension as long as ffmpeg supports it? So I was keeping it open and up to ffmpeg whether it supports the output extension. But do you think we need to limit them using `ACCEPTABLE_EXTENSIONS`?

calw20 replied:
I added the check as a bit of a bodge; it was something I was hopeing to clean up a bit later a work & wanted to seek comments on - I added it as I got an erorr from `Loom` when trying to create an `.avi` movie file. From memory the `command` list wasn't created as the conditional logic block didn't have a default else case that functioned.

ali-ramadhan replied:
Now that's a format I haven't seen in a while haha.

It's true that we don't really test for any output format besides GIF and MP4. I would say if AVI output does not work, let's open an issue about it and hopefully it'll be an easy fix.

ali-ramadhan reviewed (changes requested):
Hi @calw20 and thank you for contributing by opening this PR! I didn't know that matplotlib came with an ffmpeg executable and I think using the matplotlib ffmpeg is a great improvement, especially for users on systems with no ffmpeg (or a very old ffmpeg version).

Out of curiosity, do you know if ffmpeg versions are unique to matplotlib versions? E.g. if you have matplotlib version x.y.z then you're guaranteed to have ffmpeg version a.b.c?

I just left a couple of comments/questions to address before we merge. And thanks for tagging v0.9.2!

calw20 commented:
As far as I know matplotlib _also_ defaults to trying the `ffmpeg` command BUT though rcPrams configuration it allows you to point to a specific ffmpeg path as well. I used this method for work as I use an ffmpeg python module (its basically a wrapper to a single binary & a python method to get the path of the binary) as I use `uv` based virtual environments.

calw20 commented:
I can show some examples later next week once I'm back at work - if it helps as well. I'm super chuffed how easy it was to use `Loom` so thankyou for the quick speed up in generating video plots!

ali-ramadhan commented on matplotloom/__init__.py:
```diff
@@ -27,12 +28,15 @@ def _check_ffmpeg_availability():
         "Please install ffmpeg to use this library. "
-        "Visit https://ffmpeg.org/download.html for installation instructions.",
+        "Visit https://ffmpeg.org/download.html for installation instructions."
+        "Or configure `matplotlib.pyplot.rcParams['animation.ffmpeg_path']` to "
```
```suggestion
        "Visit https://ffmpeg.org/download.html for installation instructions. "
        "Or configure `matplotlib.pyplot.rcParams['animation.ffmpeg_path']` to "
```

ali-ramadhan commented on matplotloom/loom.py:
```diff
@@ -107,6 +116,13 @@ def __init__(
+        if ffmpeg_path is not None:
+            _ffmpeg_path = Path(ffmpeg_path)
+            if _ffmpeg_path.exists():
+                self.ffmpeg_path = str(_ffmpeg_path.absolute())
```
This is great but if the user provides an `ffmpeg_path` that doesn't exist, the code will silently fall back to the default. To alert the user and avoid confusing behavior, I think we should either produce an error and exit, or warn the user then switch to the default.

ali-ramadhan reviewed (commented):
Hey @calw20 sorry I've been meaning to revisit this PR. I left a few more comments. Just let me know if I can help with this PR.

I might make some changes to the rest of the code (nothing too drastic) but if any conflicts show up in this PR I can help resolve them.

calw20 committed:
Apply formatting suggestion from @ali-ramadhan

Co-authored-by: Ali Ramadhan <ali.hh.ramadhan@gmail.com>

calw20 committed:
Make default path more clear

calw20 committed:
Add optional path fallback & warning to ffmpeg path provider

calw20 committed:
Make the default ffmpeg path one that is sync'd across multi-processes

calw20 committed:
Fix bug where default path isn't used if an override path isn't provided

calw20 committed:
Add example of overloading matplotlibs ffmpeg rcPrams setting with a path provided by imageio_ffmpeg

calw20 commented:
Hey @ali-ramadhan, I also have to apologise, I also have been meaning to continue looking at this PR but got caught up in end-of-year work shenanigans. 

I've made a few additional commits, the crux of which are:
- I agree with [your point](https://github.com/ali-ramadhan/matplotloom/pull/39#discussion_r2765450921) of having an optional warning be thrown and have added a `enable_ffmpeg_path_fallback` option (defaulting to `True`)
- I've implemented the `DEFAULT_FFMPEG_PATH` const to be pulled from an environment variable as I ran into a bug where parallelised instances of the `Loom` object wouldn't get the correct rcPrmas settings from `matplotlib` if they are in a new process AND the rcPrams where set via assignment (ie only present in the scope of the main process)
- Added an example of using a custom ffmpeg path to [`parallel_sine_wave.py`](https://github.com/ali-ramadhan/matplotloom/pull/39/changes#diff-202367c77bf75089fd0aef7a8431b036d52f48423dd9378f068baa4f71e62047R5-R5)

Hope this helps!

ali-ramadhan commented on matplotloom/loom.py:
```diff
@@ -107,6 +127,38 @@ def __init__(
+                    )            
+                
+        # In theory this should never fail.
+        assert isinstance(self.ffmpeg_path, str), "ffmpeg path is not a string?"
```
Hmmm, I'm thinking we will probably want `self.ffmpeg_path` to always be a `Path` object?

ali-ramadhan commented on matplotloom/loom.py:
```diff
@@ -233,6 +292,8 @@ def _get_scale_filter(self) -> str:
         elif self.odd_dimension_handling == "pad":
             return "pad='if(mod(iw,2),iw+1,iw)':'if(mod(ih,2),ih+1,ih)':0:0:color=white"
+        else:
+            raise ValueError("Scale Settings Incorrect!")
```
Might be good to print the value of `self.odd_dimension_handling` and the possible values so the user can know what is wrong.

ali-ramadhan commented on examples/parallel_sine_wave.py:
```diff
@@ -2,6 +2,22 @@
+
+# Either the environ var OR matplotlib rcPram *needs* to be set BEFORE
+#  importing matplotloom
+import imageio_ffmpeg # python library that provides ffmpeg binary 
```
Nice, didn't know about this package!

ali-ramadhan reviewed (commented):
@calw20 Thanks for revisiting this PR! Nice work adding the `enable_ffmpeg_path_fallback` option and the example.

Took another look and I have a few more thoughts:

1. Good idea to add an example of using a custom ffmpeg path! But I think we should not add it to the `parallel_sine_wave.py` example. I think each example should be as simple and self-contained as possible. So I would actually suggest a completely new example: `custom_ffmpeg_sine_wave.py`? Then we can test it in CI and include it in the README and the docs.

2. I think we should get rid of `ACCEPTABLE_EXTENSIONS` and aim to support any extension by default (that ffmpeg also supports). If a particular extension causes an issue, we can open a GitHub issue and work on fixing it. I feel like ffmpeg supports a huge range of output file formats, and many of them would probably work out of the box for us here.

3. I'm a little confused by what `_LOOM_DEFAULT_ENVIRON_VAR` does and the use case for it.

4. Can we add some tests for this new functionality? Should be simple to test the logic of `ffmpeg_path` and `enable_ffmpeg_path_fallback`. Testing the new example would also be super useful. I'll try to think of others too.

5. Would also be great if we can add a section to the docs documenting the default path and the different ways you can use a custom ffmpeg path.

Apologies for the many asks! Very happy to work on some of them!

calw20 committed:
Remove Limits on Filetype

calw20 committed:
Move custom ffmpeg example to 'custom_ffmpeg_sine_wave.py'

calw20 committed:
Make \self.ffmpeg_path\ be a path object

calw20 committed:
Make \loom._get_scale_filter\ return the accepted scale options on error.

calw20 commented:
@ali-ramadhan I've managed jump on a few of the items mentioned but haven't had a chance to look at all of them.

re; your comments
1. Yeah absolutely, was me just pushing an example rather than making a correct one 😅. I've moved my example over to `custom_ffmpeg_sine_wave.py` - its a copy past of `parallel_sine_wave.py` so it could be changed to another equation for the sake of difference. I'm not too sure how to best "show off" the ability to provide an alternate default ffmpeg path however.

2.  Done! (I think) - I was being dense for a bit there and didn't follow what you meant [here](https://github.com/ali-ramadhan/matplotloom/pull/39#discussion_r2765459845) but am on the same page now.

3. The use of a global constant for the env-var is was just my personal preference to minimise "magic constants" - I figured if is something we wanted to have printed in a message and was checked more than once in code it should be a constant rather then a bunch of strings all over the place.
    - Additionally I wasn't sure what would be wanted as the env-var name so I wanted to make it as simple to change.

4. We should yeah, I've been ~~procrastinating~~ _dancing around_ writing the tests partially as I'm not that familiar with `pytest` and partially as my testing has been making graphs in a work project. Very helpful I know 😅.

5. ~I need to wrap my head around Sphinx docs but those should be added shortly~ This should be done in [this](https://github.com/ali-ramadhan/matplotloom/pull/39/changes/b5576890f3ed0371607c8680c72c8acda61dc998) commit.

calw20 committed:
Add some simple docs about providing a custom FFmpeg path

calw20 committed:
Merge pull request #1 from ali-ramadhan/main

Track Main Branch

calw20 commented:
Have confirmed tests still pass on Windows after bumping to CPython 3.14.3

ali-ramadhan commented:
Awesome @calw20! Sorry I haven't been back here in a while. Will review in the next day or two!
